File: benchmarks/gene_info/bin/make_rdf.c
#!/usr/bin/env runc
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_LINE_LENGTH 8192

void format_str_array(char* str, char* formatted_str) {
    char* token = strtok(str, "|");
    while (token != NULL) {
        strcat(formatted_str, "\"");
        strcat(formatted_str, token);
        strcat(formatted_str, "\" ,\n        ");
        token = strtok(NULL, "|");
    }
    formatted_str[strlen(formatted_str) - 11] = '\0';  // Remove trailing comma and space
}

void format_link(char* str, char* formatted_link, char* filtered_db_xref) {
    char* token = strtok(str, "|");
    while (token != NULL) {
        if (strncmp(token, "MIM:", 4) == 0) {
            strcat(formatted_link, "mim:");
            strcat(formatted_link, token + 4);
            strcat(formatted_link, " ,\n        ");
        } else if (strncmp(token, "HGNC:HGNC:", 10) == 0) {
            strcat(formatted_link, "hgnc:");
            strcat(formatted_link, token + 10);
            strcat(formatted_link, " ,\n        ");
        } else if (strncmp(token, "Ensembl:", 8) == 0) {
            strcat(formatted_link, "ensembl:");
            strcat(formatted_link, token + 8);
            strcat(formatted_link, " ,\n        ");
        } else if (strncmp(token, "miRBase:", 8) == 0) {
            strcat(formatted_link, "mirbase:");
            strcat(formatted_link, token + 8);
            strcat(formatted_link, " ,\n        ");
        } else {
            strcat(filtered_db_xref, "\"");
            strcat(filtered_db_xref, token);
            strcat(filtered_db_xref, "\", ");
        }
        token = strtok(NULL, "|");
    }
    formatted_link[strlen(formatted_link) - 11] = '\0';  // Remove trailing comma and space
    filtered_db_xref[strlen(filtered_db_xref) - 2] = '\0';  // Remove trailing comma and space
}

void filter_str(char* str, char* filtered_str) {
    char* token = strtok(str, "|");
    while (token != NULL) {
        if (strncmp(token, "MIM:", 4) != 0 && strncmp(token, "HGNC:HGNC:", 10) != 0 &&
            strncmp(token, "Ensembl:", 8) != 0 && strncmp(token, "miRBase:", 8) != 0) {
            strcat(filtered_str, "\"");
            strcat(filtered_str, token);
            strcat(filtered_str, "\", ");
        }
        token = strtok(NULL, "|");
    }
    filtered_str[strlen(filtered_str) - 2] = '\0';  // Remove trailing comma and space
}

int main() {
    char line[MAX_LINE_LENGTH];
    const char* PREFIXES = "\
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n\
@prefix dct: <http://purl.org/dc/terms/> .\n\
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .\n\
@prefix ncbigene: <http://identifiers.org/ncbigene/> .\n\
@prefix taxid: <http://identifiers.org/taxonomy/> .\n\
@prefix hgnc: <http://identifiers.org/hgnc/> .\n\
@prefix mim: <http://identifiers.org/mim/> .\n\
@prefix mirbase: <http://identifiers.org/mirbase/> .\n\
@prefix ensembl: <http://identifiers.org/ensembl/> .\n\
@prefix nuc: <http://ddbj.nig.ac.jp/ontologies/nucleotide/> .\n\
@prefix : <http://purl.org/net/orthordf/hOP/ontology#> .\n";

    printf("%s", PREFIXES);

    while (fgets(line, sizeof(line), stdin) != NULL) {
        char *newline = strchr(line, '\n');
        if (newline != NULL) {
            *newline = '\0';  // Remove trailing newline
        }

        char* token = strtok(line, "\t");
        char* field[16];
        int i = 0;
        while (token != NULL) {
            field[i] = token;
            token = strtok(NULL, "\t");
            i++;
        }

        printf("\n");
        printf("ncbigene:%s a nuc:Gene ;\n", field[1]);
        printf("    dct:identifier %s ;\n", field[1]);
        printf("    rdfs:label \"%s\" ;\n", field[2]);

        if (strcmp(field[10], "-") != 0) {
            printf("    nuc:standard_name \"%s\" ;\n", field[10]);
        }

        if (strcmp(field[4], "-") != 0) {
            char formatted_synonyms[MAX_LINE_LENGTH] = "";
            format_str_array(field[4], formatted_synonyms);
            printf("    nuc:gene_synonym %s ;\n", formatted_synonyms);
        }

        printf("    dct:description \"%s\" ;\n", field[8]);

        if (strcmp(field[13], "-") != 0) {
            char formatted_others[MAX_LINE_LENGTH] = "";
            format_str_array(field[13], formatted_others);
            printf("    dct:alternative %s ;\n", formatted_others);
        }

        char formatted_link[MAX_LINE_LENGTH] = "";
        char filtered_db_xref[MAX_LINE_LENGTH] = "";
        if (strcmp(field[5], "-") != 0) {
            format_link(field[5], formatted_link, filtered_db_xref);
            printf("    nuc:dblink %s ;\n", formatted_link);
        }

        printf("    :typeOfGene \"%s\" ;\n", field[9]);

        if (strcmp(field[12], "O") == 0) {
            printf("    :nomenclatureStatus \"official\" ;\n");
        } else if (strcmp(field[12], "I") == 0) {
            printf("    :nomenclatureStatus \"interim\" ;\n");
        }

        if (strcmp(field[11], "-") != 0) {
            printf("    :fullName \"%s\" ;\n", field[11]);
        }

        if (strlen(filtered_db_xref) > 0) {
            printf("    nuc:db_xref %s ;\n", filtered_db_xref);
        }

        if (strcmp(field[15], "-") != 0) {
            char formatted_feature_type[MAX_LINE_LENGTH] = "";
            format_str_array(field[15], formatted_feature_type);
            printf("    :featureType %s ;\n", formatted_feature_type);
        }

        printf("    :taxid taxid:%s ;\n", field[0]);
        printf("    nuc:chromosome \"%s\" ;\n", field[6]);
        printf("    nuc:map \"%s\" ;\n", field[7]);

        char* date = field[14];
        if (strlen(date) == 8) {
            char formatted_date[MAX_LINE_LENGTH] = "";
            strncpy(formatted_date, date, 4);
            strcat(formatted_date, "-");
            strncat(formatted_date, date + 4, 2);
            strcat(formatted_date, "-");
            strncat(formatted_date, date + 6, 2);
            printf("    dct:modified \"%s\"^^xsd:date .\n", formatted_date);
        }
    }

    return 0;
}

File: benchmarks/gene_info/bin/make_rdf.rb
#!/usr/bin/env ruby

input_file = ARGV[0]
unless input_file
  puts "Usage: #{File.basename($0)} <input_file>"
  exit 1
end
unless File.exist?(input_file)
  puts "Error: cannot open '#{input_file}'"
  exit 1
end

puts '@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .'
puts '@prefix dct: <http://purl.org/dc/terms/> .'
puts '@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .'
puts '@prefix ncbigene: <http://identifiers.org/ncbigene/> .'
puts '@prefix taxid: <http://identifiers.org/taxonomy/> .'
puts '@prefix hgnc: <http://identifiers.org/hgnc/> .'
puts '@prefix mim: <http://identifiers.org/mim/> .'
puts '@prefix mirbase: <http://identifiers.org/mirbase/> .'
puts '@prefix ensembl: <http://identifiers.org/ensembl/> .'
puts '@prefix nuc: <http://ddbj.nig.ac.jp/ontologies/nucleotide/> .'
puts '@prefix : <http://purl.org/net/orthordf/hOP/ontology#> .'

def format_str_array(str)
  return str.split('|')
           .map { |x| "\"#{x}\"" }
           .join(" ,\n        ")
end

def format_links(str)
  str_array = []
  str.split('|').each do |a|
    if a =~ /^MIM:(\d+)$/
      str_array << "mim:#{$1}"
    elsif a =~ /^HGNC:HGNC:(\d+)$/
      str_array << "hgnc:#{$1}"
    elsif a =~ /^Ensembl:(ENSG\d+)$/
      str_array << "ensembl:#{$1}"
    elsif a =~ /^miRBase:(MI\d+)$/
      str_array << "mirbase:#{$1}"
    end
  end
  return str_array.join(" ,\n        ")
end

def filter_str(str)
  str_array = []
  str.split('|').each do |a|
    if a =~ /^MIM:(\d+)$/
    elsif a =~ /^HGNC:HGNC:(\d+)$/
    elsif a =~ /^Ensembl:(ENSG\d+)$/
    elsif a =~ /^miRBase:(MI\d+)$/
    else
      str_array << "\"#{a}\""
    end
  end
  return str_array.join(" ,\n        ")
end

def format_date(date)
  r = date.match(/^(\d{4})(\d{2})(\d{2})$/)
  return "#{r[1]}-#{r[2]}-#{r[3]}"
end

File.open(input_file, 'r') do |file|
  header = file.readline
  file.each_line do |line|
    line = line.chomp
    fields = line.split("\t")
    puts
    puts "ncbigene:#{fields[1]} a nuc:Gene ;"
    puts "    dct:identifier #{fields[1]} ;"
    puts "    rdfs:label \"#{fields[2]}\" ;"
    if fields[10] != "-"
      puts "    nuc:standard_name \"#{fields[10]}\" ;"
    end
    if fields[4] != "-"
      synonyms = format_str_array(fields[4])
      puts "    nuc:gene_synonym #{synonyms} ;"
    end
    puts "    dct:description \"#{fields[8]}\" ;"
    if fields[13] != "-"
      alternatives = format_str_array(fields[13])
      puts "    dct:alternative #{alternatives} ;"
    end
    if fields[5] != "-"
      links = format_links(fields[5])
      if links != ""
        puts "    nuc:dblink #{links} ;"
      end
    end
    puts "    :typeOfGene \"#{fields[9]}\" ;"
    if fields[12] == "O"
      puts "    :nomenclatureStatus \"official\" ;"
    elsif fields[12] == "I"
      puts "    :nomenclatureStatus \"interim\" ;"
    end
    if fields[11] != "-"
      puts "    :fullName \"#{fields[11]}\" ;"
    end
    if fields[5] != "-"
      db_xref = filter_str(fields[5])
      if db_xref != ""
        puts "    nuc:db_xref #{db_xref} ;"
      end
    end
    if fields[15] != "-"
      feature_type = format_str_array(fields[15])
      puts "    :featureType #{feature_type} ;"
    end
    puts "    :taxid taxid:#{fields[0]} ;"
    puts "    nuc:chromosome \"#{fields[6]}\" ;"
    puts "    nuc:map \"#{fields[7]}\" ;"
    date = format_date(fields[14])
    puts "    dct:modified \"#{date}\"^^xsd:date ."
  end
end
